File: driver/tcp_driver.c
// TCP load driver ("wrk for raw TCP"): the fixed side of the client/server
// benchmarks. Plain C, one blocking thread per connection, no event loop, so
// the runtime under test is always the interesting side. Build:
//
//   cc -O2 -o tcp_driver tcp_driver.c -lpthread
//
// Modes (against the per-runtime tcp_server subjects):
//   echo   write --size bytes, read them back, --msgs times per connection;
//          --pipeline P keeps P messages in flight
//   send   stream --mb total (split across conns) to a sink server
//   recv   read --mb total (split across conns) from a source server
//
// usage: tcp_driver --mode=echo|send|recv [--host=H] [--port=P] [--conns=N]
//                   [--msgs=N] [--size=N] [--pipeline=N] [--mb=N]
#define _GNU_SOURCE
#include <arpa/inet.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <pthread.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

static const char* g_host = "127.0.0.1";
static int g_port = 18800;
static long g_conns = 1;
static long g_msgs = 100000;
static long g_size = 64;
static long g_pipeline = 1;
static long g_mb = 4096;
static enum { MODE_ECHO, MODE_SEND, MODE_RECV } g_mode = MODE_ECHO;

static pthread_barrier_t g_barrier;

static int connect_retry(void) {
    struct sockaddr_in addr = {0};
    addr.sin_family = AF_INET;
    addr.sin_port = htons(g_port);
    inet_pton(AF_INET, g_host, &addr.sin_addr);
    for (int attempt = 0; attempt < 200; attempt++) {
        int fd = socket(AF_INET, SOCK_STREAM, 0);
        if (fd < 0) return -1;
        if (connect(fd, (struct sockaddr*)&addr, sizeof(addr)) == 0) {
            int one = 1;
            setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, &one, sizeof(one));
            return fd;
        }
        close(fd);
        usleep(10000); // server may still be starting
    }
    return -1;
}

static int write_full(int fd, const char* buf, size_t len) {
    while (len > 0) {
        ssize_t n = write(fd, buf, len);
        if (n <= 0) return -1;
        buf += n;
        len -= n;
    }
    return 0;
}

static int read_full(int fd, char* buf, size_t len) {
    while (len > 0) {
        ssize_t n = read(fd, buf, len);
        if (n <= 0) return -1;
        buf += n;
        len -= n;
    }
    return 0;
}

static void* worker(void* arg) {
    long* result = arg; // 0 = ok
    *result = -1;

    int fd = connect_retry();
    if (fd < 0) {
        fprintf(stderr, "connect failed\n");
        pthread_barrier_wait(&g_barrier);
        return NULL;
    }
    char* buf = calloc(1, g_size);

    pthread_barrier_wait(&g_barrier); // start together, after all connected

    switch (g_mode) {
    case MODE_ECHO: {
        long inflight = g_pipeline < g_msgs ? g_pipeline : g_msgs;
        long sent = 0, done = 0;
        for (; sent < inflight; sent++) {
            if (write_full(fd, buf, g_size) != 0) goto out;
        }
        while (done < g_msgs) {
            if (read_full(fd, buf, g_size) != 0) goto out;
            done++;
            if (sent < g_msgs) {
                if (write_full(fd, buf, g_size) != 0) goto out;
                sent++;
            }
        }
        break;
    }
    case MODE_SEND: {
        long left = g_mb * 1024 * 1024 / g_conns;
        while (left > 0) {
            long n = left < g_size ? left : g_size;
            if (write_full(fd, buf, n) != 0) goto out;
            left -= n;
        }
        break;
    }
    case MODE_RECV: {
        long left = g_mb * 1024 * 1024 / g_conns;
        while (left > 0) {
            ssize_t n = read(fd, buf, left < g_size ? left : g_size);
            if (n <= 0) goto out;
            left -= n;
        }
        break;
    }
    }
    *result = 0;
out:
    free(buf);
    close(fd);
    return NULL;
}

int main(int argc, char** argv) {
    for (int i = 1; i < argc; i++) {
        if (!strncmp(argv[i], "--mode=", 7)) {
            const char* m = argv[i] + 7;
            if (!strcmp(m, "echo")) g_mode = MODE_ECHO;
            else if (!strcmp(m, "send")) g_mode = MODE_SEND;
            else if (!strcmp(m, "recv")) g_mode = MODE_RECV;
            else goto usage;
        } else if (!strncmp(argv[i], "--host=", 7)) g_host = argv[i] + 7;
        else if (!strncmp(argv[i], "--port=", 7)) g_port = atoi(argv[i] + 7);
        else if (!strncmp(argv[i], "--conns=", 8)) g_conns = atol(argv[i] + 8);
        else if (!strncmp(argv[i], "--msgs=", 7)) g_msgs = atol(argv[i] + 7);
        else if (!strncmp(argv[i], "--size=", 7)) g_size = atol(argv[i] + 7);
        else if (!strncmp(argv[i], "--pipeline=", 11)) g_pipeline = atol(argv[i] + 11);
        else if (!strncmp(argv[i], "--mb=", 5)) g_mb = atol(argv[i] + 5);
        else goto usage;
    }
    if (g_conns < 1 || g_size < 1 || g_pipeline < 1) goto usage;

    pthread_barrier_init(&g_barrier, NULL, g_conns + 1);

    pthread_t* threads = calloc(g_conns, sizeof(pthread_t));
    long* results = calloc(g_conns, sizeof(long));
    for (long i = 0; i < g_conns; i++) {
        pthread_create(&threads[i], NULL, worker, &results[i]);
    }

    pthread_barrier_wait(&g_barrier); // all connected
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    long failed = 0;
    for (long i = 0; i < g_conns; i++) {
        pthread_join(threads[i], NULL);
        if (results[i] != 0) failed++;
    }
    clock_gettime(CLOCK_MONOTONIC, &t1);

    double secs = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) / 1e9;
    if (failed) {
        fprintf(stderr, "%ld/%ld connections failed\n", failed, g_conns);
        return 1;
    }
    if (g_mode == MODE_ECHO) {
        long total = g_msgs * g_conns;
        printf("echo: %.3f ms, %ld msgs over %ld conns, size %ld, pipeline %ld, %.0f msgs/s\n",
               secs * 1000.0, total, g_conns, g_size, g_pipeline, total / secs);
    } else {
        double gb = g_mb * 1024.0 * 1024.0 / g_conns * g_conns / 1e9;
        printf("%s: %.3f ms, %ld MB over %ld conns, chunk %ld, %.2f GB/s\n",
               g_mode == MODE_SEND ? "send" : "recv",
               secs * 1000.0, g_mb, g_conns, g_size, gb / secs);
    }
    return 0;

usage:
    fprintf(stderr, "usage: tcp_driver --mode=echo|send|recv [--host=H] [--port=P] [--conns=N] [--msgs=N] [--size=N] [--pipeline=N] [--mb=N]\n");
    return 2;
}

File: driver/tcp_driver.go
// TCP load driver on Go: one goroutine per connection over the runtime's
// netpoller, so N connections cost ~GOMAXPROCS OS threads instead of N. A
// lower-overhead alternative to the old C driver (one blocking OS thread per
// connection), which becomes the bottleneck at high connection counts. Same
// flags, modes, and output as the C driver, so it is a drop-in replacement.
//
//	echo   write --size bytes, read them back, --msgs times per connection;
//	       --pipeline P keeps P messages in flight
//	send   stream --mb total (split across conns) to a sink server
//	recv   read --mb total (split across conns) from a source server
package main

import (
	"flag"
	"fmt"
	"io"
	"net"
	"os"
	"sync"
	"time"
)

func main() {
	mode := flag.String("mode", "echo", "echo|send|recv")
	host := flag.String("host", "127.0.0.1", "server host")
	port := flag.Int("port", 18800, "server port")
	conns := flag.Int("conns", 1, "connections")
	msgs := flag.Int("msgs", 100000, "echo messages per connection")
	size := flag.Int("size", 64, "message / chunk size")
	pipeline := flag.Int("pipeline", 1, "echo messages kept in flight")
	mb := flag.Int64("mb", 4096, "send/recv total MB")
	flag.Parse()
	if *conns < 1 || *size < 1 || *pipeline < 1 {
		fmt.Fprintln(os.Stderr, "usage: tcp_driver --mode=echo|send|recv [--host=H] [--port=P] [--conns=N] [--msgs=N] [--size=N] [--pipeline=N] [--mb=N]")
		os.Exit(2)
	}
	addr := fmt.Sprintf("%s:%d", *host, *port)

	// Two barriers: connect everyone first, then start the timed work together.
	var connected, done sync.WaitGroup
	connected.Add(*conns)
	done.Add(*conns)
	start := make(chan struct{})
	results := make([]error, *conns)

	for i := 0; i < *conns; i++ {
		go func(i int) {
			defer done.Done()
			c, err := connectRetry(addr)
			if err != nil {
				fmt.Fprintln(os.Stderr, "connect failed")
				results[i] = err
				connected.Done()
				return
			}
			defer c.Close()
			connected.Done()
			<-start
			results[i] = work(c, *mode, *msgs, *size, *pipeline, *mb, int64(*conns))
		}(i)
	}

	connected.Wait() // all connected, start together
	t0 := time.Now()
	close(start)
	done.Wait()
	secs := time.Since(t0).Seconds()

	failed := 0
	for _, e := range results {
		if e != nil {
			failed++
		}
	}
	if failed > 0 {
		fmt.Fprintf(os.Stderr, "%d/%d connections failed\n", failed, *conns)
		os.Exit(1)
	}

	if *mode == "echo" {
		total := int64(*msgs) * int64(*conns)
		fmt.Printf("echo: %.3f ms, %d msgs over %d conns, size %d, pipeline %d, %.0f msgs/s\n",
			secs*1000, total, *conns, *size, *pipeline, float64(total)/secs)
	} else {
		perConn := *mb * 1024 * 1024 / int64(*conns)
		gb := float64(perConn*int64(*conns)) / 1e9
		fmt.Printf("%s: %.3f ms, %d MB over %d conns, chunk %d, %.2f GB/s\n",
			*mode, secs*1000, *mb, *conns, *size, gb/secs)
	}
}

func connectRetry(addr string) (net.Conn, error) {
	for attempt := 0; attempt < 200; attempt++ {
		c, err := net.Dial("tcp", addr) // Go enables TCP_NODELAY by default
		if err == nil {
			return c, nil
		}
		time.Sleep(10 * time.Millisecond) // server may still be starting
	}
	return nil, fmt.Errorf("connect failed")
}

func work(c net.Conn, mode string, msgs, size, pipeline int, mb, conns int64) error {
	buf := make([]byte, size)
	switch mode {
	case "echo":
		inflight := pipeline
		if inflight > msgs {
			inflight = msgs
		}
		sent, got := 0, 0
		for ; sent < inflight; sent++ {
			if _, err := c.Write(buf); err != nil {
				return err
			}
		}
		for got < msgs {
			if _, err := io.ReadFull(c, buf); err != nil {
				return err
			}
			got++
			if sent < msgs {
				if _, err := c.Write(buf); err != nil {
					return err
				}
				sent++
			}
		}
	case "send":
		left := mb * 1024 * 1024 / conns
		for left > 0 {
			n := int64(size)
			if left < n {
				n = left
			}
			if _, err := c.Write(buf[:n]); err != nil {
				return err
			}
			left -= n
		}
	case "recv":
		left := mb * 1024 * 1024 / conns
		for left > 0 {
			n := int64(size)
			if left < n {
				n = left
			}
			m, err := c.Read(buf[:n])
			if m > 0 {
				left -= int64(m)
			}
			if err != nil {
				break
			}
		}
		if left > 0 {
			return fmt.Errorf("short recv")
		}
	default:
		return fmt.Errorf("unknown mode %q", mode)
	}
	return nil
}
